### ISSUE TITLE
Enhancement: `no-floating-promises` - don't trigger on browser event handlers

### ISSUE THREAD

**Issue by benmccann** [typescript-eslint/typescript-eslint#9340]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises/

### Description

`no-floating-promises` helps ensure that you don't have unhandled rejected promises, which is helpful when you can bubble exceptions up to an eventual try/catch. However, in the case of event handlers there's nothing you can really do. You could wrap it in a function like:
```
function handlePromiseError(fn: () => Promise<unknown>): Promise<unknown> {
  return fn().catch((error: unknown) => {
    console.error(error);
  });
}
```

But this is really no better than just handling the [`unhandledrejection`](https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event) event in the browser. Handling `unhandledrejection` here is much nicer because you only have to do it once as opposed to at every single call site.

### Fail

```TypeScript
n/a
```

### Pass

```TypeScript
element.addEventListener('click', async (event) => {
   ...
});
```

**Comment by bradzacher** [comment]
[`no-floating-promise` does not report on such code](https://typescript-eslint.io/play/#ts=5.4.5&fileType=.tsx&code=CYewxgrgtgpgdgFwHQHMYIEohAgciYGACgEokBDYYAUQDd4EAZASwGcF4YAnIgcjAA2zMAGteAGgAE5VgE84YSURj1EJSQF4AfJIDeAKElGj%2BgL4l9QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

[Did you perhaps mean `no-misused-promises`](https://typescript-eslint.io/play/#ts=5.4.5&fileType=.tsx&code=CYewxgrgtgpgdgFwHQHMYIEohAgciYGACgEokBDYYAUQDd4EAZASwGcF4YAnIgcjAA2zMAGteAGgAE5VgE84YSURj1EJSQF4AfJIDeAKElGj%2BgL4l9QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)?

**Comment by bradzacher** [comment]
Our lint rules do not encode implementation details about functions - even those as part of "standard apis". Why? A few reasons:
1) those implementation details are subject to change - which creates a maintenance burden for us
2) implementation details can be overridden by a user at any time in JS
3) attempting to analyse for those things can be difficult given TS's fairly dynamic type system

In your example you didn't define the type of `element`. Was it meant to be a DOM element (eg from `document.getElementById(xxx)`)? I assume yes.

For (3) - we need to check where the type of `addEventListener` comes from. EG we would need to determine if it is it the `EventTarget` type defined in `lib.dom.d.ts`, or is it another type. If it's another type - what do we do...? 
For example imagine that your codebase practiced [the law of demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) and instead `element` was typed as, say
```ts
interface Listenable {
  addEventListener: ( ..args.. ) => void
}
function doWork(element: Listenable) {
  element.addEventListener(...);
}
```
Then how could we tell that `element` was originally an `EventTarget`? Well we can't - so we'd error on this code 😢.

What about the case where someone does this:
```ts
const element: EventTarget = {
  addEventListener: (ev, cb) => {
    doSomeWorkBeforeTheCb();
    cb();
    doSomeWorkAfterTheCb();
  },
  // ...
};

element.addEventListener(...);
```
Well we'd NOT error on this code even though we should.

Sadly things aren't clear-cut and well-defined in TS so it's hard for us to build some classes of things because TS can't provide the guarantees you desire.

---

Additionally on top of all of that - whose to say that everyone wants to ignore this case? For example your opinion is that handling the unhandled rejection is better - but many people might disagree. It's all a matter of perspective.
